src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,808 @@
+        /// Flag to indicate if the code fix can introduce local declaration statements
+        /// that need to be moved closer to the first reference of the declared variable.
+        /// This is currently only possible for the unused value assignment fix.
+        /// </summary>
+        private static bool NeedsToMoveNewLocalDeclarationsNearReference(string diagnosticId)
+            => diagnosticId == IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId;
+
+        protected override bool IncludeDiagnosticDuringFixAll(FixAllState fixAllState, Diagnostic diagnostic)
+        {
+            return fixAllState.CodeActionEquivalenceKey == GetEquivalenceKey(diagnostic) &&
+                !IsForEachIterationVariableDiagnostic(diagnostic, fixAllState.Document);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root,
+            out string diagnosticId,
+            out UnusedValuePreference preference,
+            out bool removeAssignments)
+        {
+            diagnosticId = diagnostics[0].Id;
+            var success = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.TryGetUnusedValuePreference(diagnostics[0], out preference);
+            Debug.Assert(success);
+            removeAssignments = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostics[0]);
+#if DEBUG
+            foreach (var diagnostic in diagnostics)
+            {
+                Debug.Assert(diagnosticId == diagnostic.Id);
+                Debug.Assert(AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.TryGetUnusedValuePreference(diagnostic, out var diagnosticPreference) &&
+                             diagnosticPreference == preference);
+                Debug.Assert(removeAssignments == AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic));
+            }
+#endif
+
+            return GetDiagnosticsGroupedByMember(diagnostics, syntaxFacts, root);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root)
+            => diagnostics.GroupBy(d => syntaxFacts.GetContainingMemberDeclaration(root, d.Location.SourceSpan.Start));
+
+        private async Task<Document> PreprocessDocumentAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            // Track all the member declaration nodes that have diagnostics.
+            // We will post process all these tracked nodes after applying the fix (see "PostProcessDocumentAsync" below in this source file).
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
```

Comment:
nit: I see there are sometimes empty lines but sometimes not. I could not get why there happen in some cases. Maybe there should not be empty lines after comments.

Reply:
I like to do so for comments at the start of the method explaining something about the entire method, as opposed to first statement. I should possibly start moving them to doc comments, though I would prefer some editorconfig rule with an auto code-fixer help me do it. Leaving this as is for now as there are quite a few of them.

---
In reply to: [234790668](https://github.com/dotnet/roslyn/pull/30777#discussion_r234790668) [](ancestors = 234790668)